compiler/rustc_passes/src/dead.rs

```diff
@@ -452,15 +452,17 @@ fn has_allow_dead_code_or_lang_attr(tcx: TyCtxt<'_>, id: hir::HirId) -> bool {
     }
 
     let def_id = tcx.hir().local_def_id(id);
-    let cg_attrs = tcx.codegen_fn_attrs(def_id);
-
-    // #[used], #[no_mangle], #[export_name], etc also keeps the item alive
-    // forcefully, e.g., for placing it in a specific section.
-    if cg_attrs.contains_extern_indicator()
-        || cg_attrs.flags.contains(CodegenFnAttrFlags::USED)
-        || cg_attrs.flags.contains(CodegenFnAttrFlags::USED_LINKER)
-    {
-        return true;
+    if tcx.has_codegen_attrs(tcx.def_kind(def_id)) {
+        let cg_attrs = tcx.codegen_fn_attrs(def_id);
```

Comment:
This pattern of `if tcx.has_codegen_attrs(tcx.def_kind(def_id)) { tcx.codegen_fn_attrs(def_id) }` comes up a lot.
Can it be replaced using a wrapper in `TyCtxt` which returns `CodegenFnAttrs::EMPTY` when not applicable?  Or else: why not return `CodegenFnAttrs::EMPTY` when not applicable?

Reply:
I removed it in some of the places where it's used where its either called for functions, statics or constants by adding a function `body_codegen_attrs` to the `TyCtxt`.

Didn't want to simply return `CodegenFnAttrs::EMPTY` to better detect misuse, not 100% sure if that's worth it, but I think that the current setup isn't too bad ^^